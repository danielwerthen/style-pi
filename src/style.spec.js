const Style = require('./style');
const {
  addPrefixMethods,
  addSuffixMethods,
} = require('meta-prototype/src/utils');
const Styletron = require('styletron-engine-atomic').Server;
const StyletronAdapter = require('./adapters/styletron-adapter');

describe('Style', () => {
  class Level1 extends Style {}

  const breakpoints = {
    small: '(max-width: 768px)',
    large: '(min-width: 769px)',
  };

  addSuffixMethods(Level1, ['hover'], function(property, pseudo, value) {
    this.addModifier('pseudo', ':' + pseudo);
    return this[property](value);
  });

  addPrefixMethods(Level1, ['small', 'large'], function(
    property,
    media,
    value,
  ) {
    this.addModifier('media', breakpoints[media]);
    return this[property](value);
  });

  class Level2 extends Level1 {
    foobar() {
      return this.spread({
        fontSize: '12px',
        smallFontSize: '18px',
        margin: '21px',
      });
    }
  }

  describe('instance of level2', () => {
    it('should correctly apply foobar spread style', () => {
      const styletron = new Styletron();
      const adapter = new StyletronAdapter(styletron);
      const instance = new Level2(adapter);
      expect(instance.foobar()).toMatchSnapshot();
      expect(styletron.getCss()).toMatchSnapshot();
    });
    it('should correctly apply foobar spread style with hover suffix', () => {
      const styletron = new Styletron();
      const adapter = new StyletronAdapter(styletron);
      const instance = new Level2(adapter);
      expect(instance.foobarHover()).toMatchSnapshot();
      expect(styletron.getCss()).toMatchSnapshot();
    });
  });

  describe('instance of level1', () => {
    it('should use the media prefixes and pseudo suffixes', () => {
      const styletron = new Styletron();
      const adapter = new StyletronAdapter(styletron);
      const instance = new Level1(adapter);
      expect(instance.smallFontSizeHover('13px')).toMatchSnapshot();
      expect(styletron.getCss()).toMatchSnapshot();
    });
  });
});
