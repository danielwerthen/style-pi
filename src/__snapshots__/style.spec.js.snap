// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Style instance of level1 should use the media prefixes and pseudo suffixes 1`] = `"ae"`;

exports[`Style instance of level1 should use the media prefixes and pseudo suffixes 2`] = `"@media (max-width: 768px){.ae:hover{font-size:13px}}"`;

exports[`Style instance of level2 should correctly apply foobar spread style 1`] = `"ae af ag"`;

exports[`Style instance of level2 should correctly apply foobar spread style 2`] = `".ae{font-size:12px}.ag{margin:21px}@media (max-width: 768px){.af{font-size:18px}}"`;

exports[`Style instance of level2 should correctly apply foobar spread style with hover suffix 1`] = `"ae af ag"`;

exports[`Style instance of level2 should correctly apply foobar spread style with hover suffix 2`] = `".ae:hover{font-size:12px}@media (max-width: 768px){.af:hover{font-size:18px}.ag:hover{margin:21px}}"`;
